File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        if (s == null || s.length() == 0) return 0;
        
        int left = 0;
        int maxLen = 0;

        Map<Character, Integer> counts = new HashMap<>();

        for (int right = 0; right < s.length(); right++){
            char currentChar = s.charAt(right);
            counts.put(currentChar, counts.getOrDefault(currentChar, 0) + 1); // updating the curent value of the key

            while(counts.get(currentChar) > 1){
                char leftChar = s.charAt(left);
                counts.put(leftChar, counts.get(leftChar) - 1); // updating the curent value of the key
                left++;
            }

            maxLen = Math.max(maxLen, right-left+1);
        }

        return maxLen;
    }
}

File: Data Structures & Algorithms/is-anagram/submission-2.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        hash_map = {}

        for char in s:
            if char in hash_map:
                hash_map[char] += 1
            else:
                hash_map[char] = 1

        for char in t:
            if char not in hash_map:
                return False
            hash_map[char] -= 1
            if hash_map[char] < 0:
                return False
                
        print(hash_map) 
        return True